compiler/rustc_trait_selection/src/solve/mod.rs

```diff
@@ -313,19 +313,24 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
             return Some(ty);
         };
 
-        // We do no always define opaque types eagerly to allow non-defining uses in the defining scope.
-        if let (DefineOpaqueTypes::No, ty::AliasKind::Opaque) = (define_opaque_types, kind) {
-            if let Some(def_id) = alias.def_id.as_local() {
-                if self
-                    .unify_existing_opaque_tys(
-                        param_env,
-                        OpaqueTypeKey { def_id, args: alias.args },
-                        self.next_ty_infer(),
-                    )
-                    .is_empty()
-                {
-                    return Some(ty);
-                }
+        // We do no always define opaque types eagerly to allow non-defining uses
+        // in the defining scope. However, if we can unify this opaque to an existing
+        // opaque, then we should attempt to eagerly reveal the opaque, and we fall
+        // through.
+        if let DefineOpaqueTypes::No = define_opaque_types
+            && let ty::Opaque = kind
+            && let Some(def_id) = alias.def_id.as_local()
+            && self.can_define_opaque_ty(def_id)
```

Comment:
I guess this was previously incorrect for `Reveal::All`? :thinking:

Reply:
No, this is purely an optimization. In `Reveal::All`, we still don't have any opaques in the opaque storage.

Reply:
my thought is: if the storage is empty we don't normalize opaque types. So for reveal all we did not previously normalize opaque types here

Reply:
Oh -- that still happens then? We don't look at the `Reveal` mode in `can_define_opaque_ty`.

Reply:
We do use `DefiningAnchor::Error` in codegen, though.